HPXMLtoOpenStudio/measure.rb

```diff
@@ -2362,17 +2373,47 @@ def self.add_component_loads_output(model, conditioned_zone, liv_load_sensors, i
     end
 
+    if conditioned_zone.thermostatSetpointDualSetpoint.is_initialized
+      thermostat = conditioned_zone.thermostatSetpointDualSetpoint.get
+      htg_sch = thermostat.heatingSetpointTemperatureSchedule.get
+      clg_sch = thermostat.coolingSetpointTemperatureSchedule.get
+    end
+
+    # Sensors
+    tin_sensor = OpenStudio::Model::EnergyManagementSystemSensor.new(model, 'Zone Mean Air Temperature')
+    tin_sensor.setName("#{Constants.ObjectNameAirflow} tin s")
+    tin_sensor.setKeyName(conditioned_zone.name.to_s)
+
+    if not htg_sch.nil?
+      htg_sp_sensor = OpenStudio::Model::EnergyManagementSystemSensor.new(model, 'Schedule Value')
+      htg_sp_sensor.setName('htg sp s')
+      htg_sp_sensor.setKeyName(htg_sch.name.to_s)
+    end
+
+    if not clg_sch.nil?
+      clg_sp_sensor = OpenStudio::Model::EnergyManagementSystemSensor.new(model, 'Schedule Value')
+      clg_sp_sensor.setName('clg sp s')
+      clg_sp_sensor.setKeyName(clg_sch.name.to_s)
+    end
+
     # EMS program: Heating vs Cooling logic
     program.addLine('Set htg_mode = 0')
     program.addLine('Set clg_mode = 0')
-    program.addLine("If (#{liv_load_sensors[:htg].name} > 0)") # Assign hour to heating if heating load
+    program.addLine("If (#{cond_load_sensors[:htg].name} > 0)") # Assign hour to heating if heating load
     program.addLine("  Set htg_mode = #{total_heat_load_served}")
-    program.addLine("ElseIf (#{liv_load_sensors[:clg].name} > 0)") # Assign hour to cooling if cooling load
+    program.addLine("ElseIf (#{cond_load_sensors[:clg].name} > 0) || (#{natvent_flow_actuator.name} > 0) || (hr_whf <> 0)") # Assign hour to cooling if cooling load or natural ventilation is operating
     program.addLine("  Set clg_mode = #{total_cool_load_served}")
-    program.addLine("ElseIf (#{@clg_ssn_sensor.name} > 0)") # No load, assign hour to cooling if in cooling season definition (Note: natural ventilation & whole house fan only operate during the cooling season)
-    program.addLine("  Set clg_mode = #{total_cool_load_served}")
-    program.addLine('Else') # No load, assign hour to heating if not in cooling season definition
-    program.addLine("  Set htg_mode = #{total_heat_load_served}")
+    program.addLine('Else') # No load, assign hour to cooling if temperature closer to cooling setpoint, otherwise assign hour to heating
+    if (not htg_sp_sensor.nil?) && (not clg_sp_sensor.nil?)
+      program.addLine("  Set Tmid_setpoint = (#{htg_sp_sensor.name} + #{clg_sp_sensor.name}) / 2") # Average of heating/cooling setpoints
+    else
+      program.addLine("  Set Tmid_setpoint = #{UnitConversions.convert(73.0, 'F', 'C')}") # Assumption when no HVAC system
+    end
+    program.addLine("  If #{tin_sensor.name} > Tmid_setpoint")
+    program.addLine("    Set clg_mode = #{total_cool_load_served}")
+    program.addLine('  Else')
+    program.addLine("    Set htg_mode = #{total_heat_load_served}")
+    program.addLine('  EndIf')
```

Comment:
Seems like this approach wouldn't handle well the situation where heating and cooling setpoints are the same. In that case, it would always choose heating mode.

Reply:
Though I just realized this code only gets hit when the temperature is floating (fallback to when there's no heating or cooling load), and you should never be floating if the heating and cooling setpoints are the same.